8324241: Always record evol_method deps to avoid excessive method flushing

simonis opened the pull request:
Currently we don't record dependencies on redefined methods (i.e. `evol_method` dependencies) in JIT compiled methods if none of the `can_redefine_classes`, `can_retransform_classes` or `can_generate_breakpoint_events` JVMTI capabalities is set. This means that if a JVMTI agent which requests one of these capabilities is dynamically attached, all the methods which have been JIT compiled until that point, will be marked for deoptimization and flushed from the code cache. For large, warmed-up applications this mean deoptimization and instant recompilation of thousands if not then-thousands of methods, which can lead to dramatic performance/latency drop-downs for several minutes.

One could argue that dynamic agent attach is now deprecated anyway (see [JEP 451: Prepare to Disallow the Dynamic Loading of Agents](https://openjdk.org/jeps/451)) and this problem could be solved by making the recording of `evol_method` dependencies dependent on the new `-XX:+EnableDynamicAgentLoading` flag isntead of the concrete JVMTI capabilities (because the presence of the flag indicates that an agent will be loaded eventually).

But there a single, however important exception to this rule and that's JFR. JFR is advertised as low overhead profiler which can be enabled in production at any time. However, when JFR is started dynamically (e.g. through JCMD or JMX) it will silently load a HotSpot internl JVMTI agent which requests the `can_retransform_classes` and retransforms some classes. This will inevitably trigger the deoptimization of all compiled methods as described above.

I'd therefor like to propose to *always* and unconditionally record `evol_method` dependencies in JIT compiled code by exporting the relevant properties right at startup in `init_globals()`:
```c++
 jint init_globals() {
   management_init();
   JvmtiExport::initialize_oop_storage();
+#if INCLUDE_JVMTI
+  JvmtiExport::set_can_hotswap_or_post_breakpoint(true);
+  JvmtiExport::set_all_dependencies_are_recorded(true);
+#endif
```

My measurements indicate that the overhead of doing so is minimal (around 1% increase of nmethod size) and justifies the benefit. E.g. a Spring Petclinic application started with `-Xbatch -XX:+UnlockDiagnosticVMOptions -XX:+LogCompilation` compiles about ~11500 methods (~9000 with C1 and ~2500 with C2) resulting in an aggregated nmethod size of around ~40bm. Additionally recording `evol_method` dependencies only increases this size be about 400kb. The ration remains about the same if we run the same experiment with only C1 or only C2.

Instead of unconditionally recording `evol_method` dependencies we could guard the recording by a new flag. But this would only make sense if that flag would be on by default and I don't know if such a flag is justified just for the rare (or non-existent?) cases where somebody wants to disable the recording of the dependencies.

Another alternative would be to remove the `set_can_hotswap_or_post_breakpoint()`/`set_all_dependencies_are_recorded()` from `JvmtiExport` altogether (and always record the dependencies). It is currently only used at a few call sites in the C1/C2 compiler (`set_can_hotswap_or_post_breakpoint()`) and once when redefining classes (`set_all_dependencies_are_recorded()`).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324241](https://bugs.openjdk.org/browse/JDK-8324241): Always record evol_method deps to avoid excessive method flushing (**Enhancement** - P4)


### Reviewers
 * [Evgeny Astigeevich](https://openjdk.org/census#eastigeevich) (@eastig - Committer) ⚠️ Review applies to [7b750da5](https://git.openjdk.org/jdk/pull/17509/files/7b750da5bad05573151c80f7e81ceeaab456bb4d)
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**) ⚠️ Review applies to [95db4a72](https://git.openjdk.org/jdk/pull/17509/files/95db4a7218be14212d11d12f59f1d0c847fa48ea)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [6d3e24ab](https://git.openjdk.org/jdk/pull/17509/files/6d3e24ab152b3f7b2e906f9e7f0b883a99b45147)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [29966635](https://git.openjdk.org/jdk/pull/17509/files/2996663532733d90e949afae431a093923845b1f)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [29966635](https://git.openjdk.org/jdk/pull/17509/files/2996663532733d90e949afae431a093923845b1f)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17509/head:pull/17509` \
`$ git checkout pull/17509`

Update a local copy of the PR: \
`$ git checkout pull/17509` \
`$ git pull https://git.openjdk.org/jdk.git pull/17509/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17509`

View PR using the GUI difftool: \
`$ git pr show -t 17509`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17509.diff">https://git.openjdk.org/jdk/pull/17509.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17509#issuecomment-1902251150)

coleenp commented:
/label add serviceability
/label add hotspot-compiler

eastig reviewed (approved):
lgtm

shipilev commented:
> Instead of unconditionally recording evol_method dependencies we could guard the recording by a new flag. But this would only make sense if that flag would be on by default and I don't know if such a flag is justified just for the rare (or non-existent?) cases where somebody wants to disable the recording of the dependencies.

I think introducing a diagnostic flag is sensible here. If we figure out much later that this solution comes with some other (worse) problems, the diagnostic flag gives us the options to: a) clearly point at this addition as the culprit; b) have the easily deployable solution to restore the original behavior.

For the change itself, we need to amend the comment near `VM_RedefineClasses::flush_dependent_code` definition that talks about this peculiar behavior, which now changes.

eastig commented:
@shipilev I filed https://bugs.openjdk.org/browse/JDK-8324318 which is related to this change.

coleenp commented:
I support adding this as a diagnostic option, therefore you won't have to make the change in VM_RedefineClasses::flush_dependent_code.

simonis commented:
Thanks everybody for looking at this. I've now guarded the feature with a diagnostic command, updated the source code comments around `VM_RedefineClasses::flush_dependent_code` and added an assertion for the new flag.

eastig commented on src/hotspot/share/runtime/globals.hpp:
```diff
@@ -2009,6 +2009,11 @@ const int ObjectAlignmentInBytes = 8;
   product(bool, ProfileExceptionHandlers, true,                             \
           "Profile exception handlers")                                     \
+                                                                            \
+  product(bool, AlwaysRecordEvolDependencies, true, DIAGNOSTIC,             \
```
As we record all dependencies not only evol_method ones, should we name it just: `AlwaysRecordDependencies`?

rose00 replied:
That’s not exactly right either.  `RecordAllDependencies` would be more like it.  Because:

  - We might record some dependencies that we know we need, and leave others out.
  - Or, we might record all dependencies, even ones we think we might not need.

(But we will need them all if somebody turns on JFR.)

I like this change.  Having a diagnostic switch means we can do a rough triage if something seems to go wrong with this change, down the road.

coleenp replied:
No, because you want to turn on/off evol_method independently of the other dependencies that the compiler is recording.

eastig commented on src/hotspot/share/runtime/globals.hpp:
```diff
@@ -2009,6 +2009,11 @@ const int ObjectAlignmentInBytes = 8;
           "Profile exception handlers")                                     \
+                                                                            \
+  product(bool, AlwaysRecordEvolDependencies, true, DIAGNOSTIC,             \
+                "Unconditionally record method dependencies on class "      \
```
"... record compiled method dependencies ..."?

rose00 replied:
(yes, “compiled methods” or even “nmethods”, or “methods in code cache”)

simonis replied:
Thanks.  I went with "nmethods" because it's the shortest alternative :)

coleenp reviewed (approved):
This looks good.

dean-long commented on src/hotspot/share/runtime/init.cpp:
```diff
@@ -115,6 +115,12 @@ void vm_init_globals() {
+#if INCLUDE_JVMTI
+  if (AlwaysRecordEvolDependencies) {
+    JvmtiExport::set_can_hotswap_or_post_breakpoint(true);
+    JvmtiExport::set_all_dependencies_are_recorded(true);
```
I think the check for AlwaysRecordEvolDependencies needs to be moved into set_can_hotswap_or_post_breakpoint and set_all_dependencies_are_recorded, otherwise don't we risk the value being accidentally reset to false when set_can_hotswap_or_post_breakpoint() is called again by JvmtiManageCapabilities::update()?

dean-long replied:
An assert works for me.  Thanks.

simonis replied:
Added assert in `JvmtiExport::set_can_hotswap_or_post_breakpoint()`. Thanks.

shipilev commented on src/hotspot/share/prims/jvmtiRedefineClasses.cpp:
```diff
@@ -4060,21 +4060,22 @@ void VM_RedefineClasses::transfer_old_native_function_registrations(InstanceKlas
 
 void VM_RedefineClasses::flush_dependent_code() {
   assert(SafepointSynchronize::is_at_safepoint(), "sanity check");
+  assert(AlwaysRecordEvolDependencies ? JvmtiExport::all_dependencies_are_recorded() : true, "sanity check");
```
This is just "assert all dependencies are recorded, unless we specifically requested not to do so", right?

```
assert(JvmtiExport::all_dependencies_are_recorded() || !AlwaysRecordEvolDependencies, "sanity check");
```

eastig reviewed (approved):
lgtm

iwanowww commented:
I support keeping the logic under a flag. I have some concerns about unconditionally turning it on. 

I expect to see significantly higher footprint overhead when an application has plenty of tiny methods and deep inlining trees. And java.lang.invoke implementation pushes it even further (with arbitrarily deep MethodHandle trees and unconditional inlining through them), so heavy users of MethodHandle API should experience higher overheads.

I suggest to make the flag experimental. Once JFR implementation is improved, it can be superseded by `-XX:+EnableDynamicAgentLoading` check.

simonis commented:
@iwanowww , @shipilev , @dean-long any more comments or concerns? Otherwise I'd like to finish this.

shipilev commented on src/hotspot/share/runtime/globals.hpp:
```diff
@@ -2009,6 +2009,11 @@ const int ObjectAlignmentInBytes = 8;
           "Profile exception handlers")                                     \
+                                                                            \
+  product(bool, AlwaysRecordEvolDependencies, true, EXPERIMENTAL,           \
+                "Unconditionally record nmethod dependencies on class "     \
+                "rewriting/transformation independently of the JVMTI "      \
+                " can_{retransform/redefine}_classes capabilities.")        \
```
Probably match the indenting of previous delcarations?

simonis replied:
Done

shipilev commented on src/hotspot/share/prims/jvmtiExport.hpp:
```diff
@@ -149,7 +149,15 @@ class JvmtiExport : public AllStatic {
+    // reset so we have to ensure that evol dependencies are always
+    // recorded from that point on.
+    assert(!_can_hotswap_or_post_breakpoint || on, "sanity check");
+    _can_hotswap_or_post_breakpoint = (on != 0);
```
Pre-existing: wild that this code checks `!= 0` against `bool`, when it could have just used the bool directly, like the new assert does. I see no reason to keep `!= 0` here.

simonis replied:
I wondered myself about this strange style. But it's there since the first OpenJDK commit (probably it was written by a C programmer where `bool` isn't a builtin type and he wanted to make sure that all non-zero values get mapped to 1).

But as this style is used throughout the entire file, I prefer to keep it as is and not change a single occurrence only.

shipilev reviewed (approved):
I am good with these changes, only a few stylistic nits.

I am on the fence where the default for this flag should stand. The 1% loss in nmethod size is probably okay, given that we gained as much with denser code improvements like [JDK-8319406](https://bugs.openjdk.org/browse/JDK-8319406). Maybe there are some tricks in dependency encoding that would drive this down even more.

simonis commented:
/integrate

RacerZ-fighting commented:
Will this feature be incorporated into a lower version of JDK, like JDK 8?

eastig commented:
@RacerZ-fighting 
> Will this feature be incorporated into a lower version of JDK, like JDK 8?

I don't see JDK 8 or 11 in affected versions of JDK-8324241, only 17 and 21.

RacerZ-fighting commented:
I've noticed that the evol_method dependency is closely related to inline optimization. I'm curious if there is a way to extract evol_method information from the results of inline optimization (perhaps nmethod)?

leomao10 commented:
Hi there, we are seeing this issue when we run JFR on our services under load, we see a large spike of CPU after JFR is triggered, which cause 500 errors in our service. We are currently using corretto-17 in our service.

Wondering this fix get back ported to JDK 17? As I can't find this change mentioned in [JDK update](https://wiki.openjdk.org/display/JDKUpdates/Archived+Releases) or in [jdk17u tag compare](https://github.com/openjdk/jdk17u/compare/jdk-17.0.9+9...jdk-17.0.13+1)

Also, wondering if there is a walk around for this issue if the PR is not back ported to Java 17. `XX:+EnableDynamicAgentLoading` seems to only supported in Java 21, so that wouldn't help for now

simonis commented:
> Hi there, we are seeing this issue when we run JFR on our services under load, we see a large spike of CPU after JFR is triggered, which cause 500 errors in our service. We are currently using corretto-17 in our service.
> 
> Wondering this fix get back ported to JDK 17? As I can't find this change mentioned in [JDK update](https://wiki.openjdk.org/display/JDKUpdates/Archived+Releases) or in [jdk17u tag compare](https://github.com/openjdk/jdk17u/compare/jdk-17.0.9+9...jdk-17.0.13+1)
> 
> Also, wondering if there is a walk around for this issue if the PR is not back ported to Java 17. `XX:+EnableDynamicAgentLoading` seems to only supported in Java 21, so that wouldn't help for now

@leomao10, I'm not sure if this change will ever be downported to older releases like JDK 21 or even JDK 17. I personally consider it low risk, but there have been reports of performance regressions in some cases (e.g. [JDK-8336805](https://bugs.openjdk.org/browse/JDK-8336805)). I couldn't reproduce them, but I can image that they will make the maintainers of LTS releases even more cautious.

The easiest way to workaround this issue in JDK 17 would be to set the `can_redefine_classes`, `can_retransform_classes` or `can_generate_breakpoint_events` JVMTI capabilities at startup or early in the lifetime of the JVM. There are several ways how you could do this.
- trigger JFR right after startup. This will still invalidate all the JIT compiled methods but if you do this early enough there won'T be many of them. After you've triggered JFR for the first time, the corresponding JVMTI capabilities will be set and all dependencies will be recorded automatically so any subsequent JFR invocation won't suffer from a performance degradation any more.
- attach any other JVMTI agent like for example [async profiler](https://github.com/async-profiler/async-profiler) which requests the corresponding JVMTI capabilities at startup.
- write your own, trivial JVMTI agent which merely requests the corresponding JVMTI capabilities and attach it at startup with `agentpath:jvmtiAgent.so`. The agent can be as simple as:
```cpp
/*

g++ -fPIC -shared -I $JAVA_HOME/include/ -I $JAVA_HOME/inlude/linux -o jvmtiAgent.so jvmtiAgent.cpp
 */

#include <jvmti.h>
#include <stdio.h>
#include <string.h>

extern "C"
JNIEXPORT jint JNICALL Agent_OnLoad(JavaVM* jvm, char* options, void* reserved) {
  jvmtiEnv* jvmti = NULL;
  jvmtiCapabilities capa;
  jvmtiError error;

  jint result = jvm->GetEnv((void**) &jvmti, JVMTI_VERSION_1_1);
  if (result != JNI_OK) {
    fprintf(stderr, "Can't access JVMTI!\n");
    return JNI_ERR;
  }

  memset(&capa, 0, sizeof(jvmtiCapabilities));
  capa.can_redefine_classes = 1;
  capa.can_retransform_classes = 1;
  capa.can_generate_breakpoint_events = 1;
  if (jvmti->AddCapabilities(&capa) != JVMTI_ERROR_NONE) {
    fprintf(stderr, "Can't set capabilities!\n");
    return JNI_ERR;
  } else {
    fprintf(stdout, "Added capabilities!\n");
  }
  return JNI_OK;
}
```

leomao10 commented:
That is awesome, thanks for the response and workarounds @simonis  ❤️
